0000-frontend-development-infrastructure-and-workflow-changes.md

```diff
@@ -0,0 +1,73 @@
+
+The frontend ecosystem has expanded dramatically with the addition of many libraries and frameworks, as well as developer productivity tools, to meet these challenges. Reactjs is among these additions.
+
+Reactjs is a view library that falls under the umbrella MV* libraries and frameworks. The common thread among these frameworks is the separation of the management of state (model) from the process of rendering HTML (view). The benefits are similar to those provided by server side MVC, such as Rails.
+
+Applying an MV* approach would facilitate enforcement of best practices and coding standards and make the code produced more of robust, maintainable and extensible. It must be noted however that, any solution must 'work well' with the current foreman Ruby/Rails code base and plugins, as there is no intention to abandon it.
+
+The tools in the MV* space are often classified as being opinionated or unopinionated with respect to the amount of flexibility they allow the developer. This often maps to the breadth of the issues the library or framework attempts to deal with.
+
+Reactjs is a component-based rendering engine. As such, it is both narrow in scope and unopinionated. This will facilitate integration with Rails with minimum side effects on the existing Ruby/Rails code base compared to the opinionated alternatives.
+
+React components demonstrate a high level of encapsulation and support dependency injection. As a result, the components are very testable. The implementation of the recommendations in this RFC will be accompanied by the addition of linting and testing to the client side development process.
+
+# Detailed design
+[design]: #detailed-design
+
+Reactjs uses javascript to compose components and render HTML. Data is passed to the component via props, which are immutable. Component state is mutable and managed by the component. When state changes the component is rerendered.
+
+Note: for those familiar with Angular 1, React components are similar to Angular directives.
+
+The [Fixes #13424 - WIP c3 patternfly react implementation PR] (https://github.com/theforeman/foreman/pull/3603) provides a POC for integrating Reactjs with foreman.
+
+Top level components added to an HTML page (mounted) with the `ReactDOM` package `render` method. This is the point at which the Rails view passes the baton to Reactjs.
+
+The paradigm for using React components in Rails views implemented in the POC is as follows:
+
+>The `mount_react_component` helper in the `ReactjsHelper` module is used in the Rails view to which the component is to be added. The helper has 3 paramaters: a string identifier for the specific component to be mounted, a jquery selector that is used to find the DOM element into which to inject the component and, optionally, initial data from the controller in JSON format. See *'app/views/statistics/index.html.erb'* in the POC for an example.
+
```

Comment:
you can write it so it includes syntax highlight(use "```erb")
```erb
<div id="statistics"></div>

<%= mount_react_component('StatisticsChartsList', '#statistics', @data)%>
```